app/scripts/lib/crypto/recovery-keys.js

```diff
@@ -0,0 +1,175 @@
+      });
+    },
+
+    /**
+     * Creates the encryption keys from the uid and recovery key. The object contains
+     * `recoveryKeyId` which can be used to retrieve the recovery data and
+     * `recoveryEnc` which can be used to decrypt the data.
+     *
+     * @param {String} uid Uid of user
+     * @param {String} recoveryKey Base32 encoded string
+     * @returns {Promise} A promise that will be fulfilled with the `recoveryKeyId` and `recoveryEnc`
+     */
+    getRecoveryEncryption: function (uid, recoveryKey) {
+      return Promise.resolve()
+        .then(() => {
+          if (! uid) {
+            throw new Error('Missing uid');
+          }
+
+          if (! recoveryKey) {
+            throw new Error('Missing recoveryKey');
+          }
+
+          const km = sjcl.codec.base32.toBits(recoveryKey);
+          const salt = sjcl.codec.hex.toBits(uid);
+
+          const info1 = sjcl.codec.utf8String.toBits('fxa recovery fingerprint');
+          const info2 = sjcl.codec.utf8String.toBits('fxa recovery encrypt key');
+
+          const recoveryKeyIdRaw = sjcl.misc.hkdf(km, 16 * 8, salt, info1);
+          const recoveryKeyEncRaw = sjcl.misc.hkdf(km, 32 * 8, salt, info2);
+
+          // Convert to hex strings for readability
+          const recoveryKeyId = sjcl.codec.hex.fromBits(recoveryKeyIdRaw);
+          const recoveryEnc = sjcl.codec.hex.fromBits(recoveryKeyEncRaw);
+
+          return {
+            recoveryEnc,
+            recoveryKeyId
+          };
+        });
+    },
+
+    /**
+     * Creates an encrypted bundle for the recovery data provided using
+     * the recovery key id and recovery encryption.
+     *
+     * This encryption is based on the protocol described in
+     *
+     * https://github.com/mozilla/fxa-auth-server/blob/master/docs/recovery_keys.md
+     *
+     * @param {String} recoveryKeyId Recovery key id used in encryption
+     * @param {String} recoveryEnc Recovery encryption used to encrypt bundle
+     * @param {Object} recoveryData Data to encrypt
+     * @param {Object} options
+     * @returns {Promise} A promise that will be fulfilled with the encrypted recoveryData
+     */
+    createRecoveryBundle: function (recoveryKeyId, recoveryEnc, recoveryData, options = {}) {
+      return importFxaCryptoDeriver().then((fxaCryptoDeriver) => {
+        const jose = fxaCryptoDeriver.jose;
+
+        if (! recoveryKeyId) {
+          throw new Error('Missing recoveryKeyId');
+        }
+
+        if (! recoveryEnc) {
+          throw new Error('Missing recoveryEnc');
+        }
+
+        const keyOptions = {
+          alg: 'A256GCM',
+          k: jose.util.base64url.encode(recoveryEnc, 'hex'),
+          kty: 'oct'
+        };
+
+        return jose.JWK.asKey(keyOptions)
+          .then(function (key) {
+            const recipient = {
+              header: {
+                alg: 'dir',
+                enc: 'A256GCM',
+                kid: recoveryKeyId
+              },
+              key
```

Comment:
obscure feature:  If you include `reference: null` in the recipient `object`, the resulting JWE header will not include a `kid`.

Personally, I think including the `kid` is still worthwhile for additional sanity checks, but only if we do them.

Reply:
> Personally, I think including the kid is still worthwhile for additional sanity checks

I agree, and I'll take a look at updating the python version to include it as well.